공지사항 댓글 목록 조회 API 개발

zlzzlzz2l opened the pull request:
## 확인사항

- [x] 테스트 코드를 작성하셨나요?
- [x] 모든 테스트 코드가 정상적으로 실행되나요?
- [x] 포스트맨으로 API 테스트를 진행하셨나요?

## 어떤 이유로 PR을 하셨나요?

- [ ] common
- [x] feature, util
- [ ] refactor
- [ ] bug-fix
- [ ] 기타 (아래 내용에 자세한 내용을 기입해주세요)

## PR 내용:

<!-- 무엇을 한 것인지 설명을 쓸 것, 글 or 글 + 사진 가능 -->

### 공지사항 댓글 조회 성공
![image](https://user-images.githubusercontent.com/48669011/170805468-1eda6df5-d6c9-4fe7-880a-b6562226b2e4.png)

### 공지사항 댓글 조회 실패 - 공지사항 데이터가 존재하지 않을 때
![image](https://user-images.githubusercontent.com/48669011/170805492-236dc0c3-88b5-4e11-b89f-defcf79f9a6d.png)

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 컨트롤러 댓글 목록 조회 추가

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 서비스 댓글 조회 선언부 추가

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 서비스 댓글 목록 구현 추가

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 응답 DTO 추가

댓글 목록을 불러오기 위해 추가하였습니다.

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 레포지토리 커스텀 추가

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 커스텀 레포지토리 선언부 추가

공지사항 식별자를 이용하여 댓글 목록을 불러오기 위해 추가하였습니다.

zlzzlzz2l committed:
#64 Feat: 공지사항 댓글 커스텀 레포지토리 구현 추가

zlzzlzz2l committed:
#64 Test: 공지사항 댓글 서비스 테스트 추가

공지사항 댓글 조회에 성공했을 때의 테스트입니다.

zlzzlzz2l committed:
#64 Test: 공지사항 댓글 서비스 테스트 추가

공지사항 댓글 조회에 실패했을 때의 테스트입니다.

zlzzlzz2l committed:
#64 Test: 공지사항 댓글 레포지토리 테스트 추가

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/service/NoticeCommentService.java:
```diff
@@ -3,9 +3,13 @@
 
     NoticeComment register(NoticeCommentDto.RegisterDto registerDto, Long memberSeq);
 
+    List<NoticeCommentDto.Detail> getDetail(Long noticeSeq);
```
메서드명을 좀 더 명확하게 바꾸는 게 어떨까요?? 댓글 목록을 반환하는 메서드니까요.
`comments()`
`commentList()`
`getComments()`
`getCommentList()`

zlzzlzz2l replied:
넵 getComments()로 반영하겠습니다!

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/service/NoticeCommentServiceImpl.java:
```diff
@@ -40,6 +42,17 @@ public NoticeComment register(NoticeCommentDto.RegisterDto registerDto, Long mem
 
+    @Override
+    public List<NoticeCommentDto.Detail> getDetail(Long noticeSeq) {
+        noticeRepository.findBySeq(noticeSeq)
+                .orElseThrow(() -> new IllegalArgumentException(ErrorMessage.NULL_NOTICE));
```
공지사항 데이터가 있는지 확인하는 거라면 `exists()` 메서드로 바꾸는 것은 어떤가요?

zlzzlzz2l replied:
넵 반영하였습니다.

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/service/NoticeCommentServiceImpl.java:
```diff
@@ -40,6 +42,17 @@ public NoticeComment register(NoticeCommentDto.RegisterDto registerDto, Long mem
+        noticeRepository.findBySeq(noticeSeq)
+                .orElseThrow(() -> new IllegalArgumentException(ErrorMessage.NULL_NOTICE));
+
+        List<NoticeCommentDto.Detail> noticeComments = noticeCommentRepository.findCommentsByNoticeSeq(noticeSeq)
+                .orElse(null);
```
아래 방식은 어떠세요?

```java
@Override
public List<NoticeCommentDto.Detail> getDetail(Long noticeSeq) {
    Notice findNotice = noticeRepository.findBySeq(noticeSeq)
                .orElseThrow(() -> new IllegalArgumentException(ErrorMessage.NULL_NOTICE));

    List<NoticeComment> comments = findNotice.getComments();
    List<NoticeCommentDto.Detail> noticeComments = new ArrayList();
    for (NoticeComment comment : comments) {
        if (!comment.isDeleted) {
            noticeComments.add(댓글DTO);
        }
    }

    return noticeComments;
}
```

zlzzlzz2l replied:
넵 반영하였습니다.

1. 조건 분기 이용하여 공지사항 데이터 여부 판단
2. 공지사항 데이터 존재 시, 댓글 DTO 생성
    a. 공지사항 데이터가 존재하기 때문에 공지사항 식별자를 이용하여 댓글을 가져오는 부분은 `get()`을 사용하였습니다.
4. 공지사항 데이터 존재하지 않을 시, exception 발생

위와 같은 흐름으로 작성하였습니다.

단, 경석님의 리뷰 중, `if (!comment.isDeleted) { }`는 반영하지 않았습니다.
그 이유는 `findBySeq()`를 실행할 때, isDeleted 조건이 붙어 `False`인 데이터를 식별해서 가져오기 때문입니다.

0xe82de replied:
말씀하신 `NoticeRepository` 클래스의 `findBySeq()` 메서드는 `Notice` 데이터를 가져오는 메서드인데.. `NoticeComment` 데이터의 `isDeleted` 조건은 따로 체크해줘야 하지 않나요?

zlzzlzz2l replied:
아하😦.. 그러네요! 제가 `NoticeRepository`와 `NoticeCommentRepository`를 착각했던 것 같아요!

반영했습니다!

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/repository/NoticeCommentRepositoryImpl.java:
```diff
@@ -0,0 +1,42 @@
+                        noticeComment.writer.nickname
+                ))
+                .from(noticeComment)
+                .where(noticeComment.notice.seq.eq(noticeSeq).and(noticeComment.isDeleted.isFalse()))
```
querydsl where 절 안에서 콤마로 분리하면 and 조건으로 묶이더라구요. 아래처럼 코드를 작성하면 읽기에 조금 더 쉬운 것 같은데.. 어떠세요?

```java
.where(
    noticeComment.notice.seq.eq(noticeSeq),
    noticeComment.isDeleted.isFalse()
)
```

zlzzlzz2l replied:
넵 반영했습니다!

0xe82de reviewed (commented):
고생하셨습니다. 코멘트 확인해주세요~!

zlzzlzz2l committed:
#64 Refactor: 메서드명 수정

getDetail() -> getComments()로 변경하였습니다.

zlzzlzz2l committed:
#64 Feat: 공지사항 레포지토리 메서드 선언부 추가

- 공지사항 존재 여부를 판단하기 위해 추가하였습니다.

zlzzlzz2l committed:
#64 Refactor: 공지사항 댓글 서비스 댓글 조회 수정

1. 공지사항 데이터 존재 여부 확인
2. 조건 분기를 이용한 데이터 추가 및 exception
    2-1. 공지사항 데이터가 존재하지 않을 경우 exception 발생
    2-2. 공지사항 데이터가 존재할 경우 댓글 DTO 생성

zlzzlzz2l committed:
#64 Refactor: 공지사항 댓글 커스텀 레포지토리 메서드 수정

0xe82de commented on src/main/java/com/fourfourfour/damoa/domain/notice/service/NoticeCommentServiceImpl.java:
```diff
@@ -44,11 +45,31 @@ public NoticeComment register(NoticeCommentDto.RegisterDto registerDto, Long mem
+                    );
+                }
+            }
+        }
```
56번 라인에서 `optional` 로 `null` 처리한다면 48~55번 라인의 코드는 없어도 괜찮을 것 같습니다!

zlzzlzz2l replied:
넵 반영했습니다!

zlzzlzz2l committed:
#64 Refactor: 공지사항 댓글 서비스 댓글 조회 수정

exists 메서드 대신 Optional을 이용하여 null 처리하였습니다.

zlzzlzz2l committed:
#64 Refactor: 공지사항 레포지토리 메서드 선언부 수정

불필요한 메서드를 삭제하였습니다.

0xe82de reviewed (approved):
고생하셨습니다~!
